### PR TITLE
SCW-309: Added getSigner

### PR DESCRIPTION
**Pull request by jordiamez**
## Description
getSigner("viem") using a custom bridge

## Test plan

Use
```
        // Get Block Number
        const blockNumber = await walletClient.getBlockNumber();
        console.log(`Current Block Number: ${blockNumber}`);

        // Sign a Message
        const message = 'hello world';
        const signature = await walletClient.signMessage({ message });
        console.log(`Signed Message: ${signature}`);

        // Verify the Signature
        const signerAddress = ethers.utils.verifyMessage(message, signature);
        console.log(`Signer Address: ${signerAddress}`);
        
        // Get Balance
        const address = '0x11d84BfB7243aB27DB3F4f2dcD1fD6d3Cc6f5E44'; 
        const balance = await walletClient.getBalance({ address });
        console.log(`Balance of ${address}: ${ethers.utils.formatEther(balance)} MATIC`);

        // Get Addresses
        const accounts = await walletClient.getAddresses();
        console.log(`Wallet Addresses: ${accounts}`);
```
On the demo to test it:
Captura de pantalla 2024-01-26 a las 17 20 42

About reservoir :
```

   const client = createClient({
      chains: [{
        ...reservoirChains.mumbai,
        active: true,
      }],
    });
   
    const walletClient = await wallet.getSigner("viem");
    
    const response =  await client.actions.buyToken({ 
      items: [{ token: "0x2953399124f0cbb46d2cbacd8a89cf0599974963:95090369944410240333816311102553893597383031364012304288034229396360438753040"}],
      wallet: walletClient,
      onProgress: (steps: Execute['steps']) => {
        console.log(steps)
      }
    })
```

Result: 
Captura de pantalla 2024-01-29 a las 16 26 33

polygonscan:

https://mumbai.polygonscan.com/address/0xbcCA9A2FEE350c9CB96CdA4a149d9912bcF1e891#nfttransfers

### CODE REVIEW

**Comment by jordiamez** [comment]
Working with reservoir sdk

File: packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+               }
+
+               const message = params![1];
+               const prefixedMessage = `\x19Ethereum Signed Message:\n${message.length}${message}`;
+               const hashedMessage = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(prefixedMessage));
+               return await this.signer.signMessage(ethers.utils.arrayify(hashedMessage));
+           }
+
+            case "eth_sendTransaction": {
+                if (!this.signer) {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts:144]:
```suggestion
                if (this.signer == null) {
```

File: packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                const address = await this.signer.getAddress();
+                if (address !== ethers.utils.getAddress(params![0])) {
+                    throw new Error('Error getting address');
+                }
+
+                return this.signer.signMessage(ethers.utils.arrayify(params![1]));
+            }
+
+            case "eth_signTransaction": {
+                if (!this.signer) {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts:115]:
```suggestion
                if (this.signer == null) {
```

File: packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            }
+            case "eth_getTransactionByHash": {
+                return await this.provider.getTransaction(params![0]);
+            }
+            case "eth_getTransactionReceipt": {
+                return await this.provider.getTransactionReceipt(params![0]);
+            }
+
+            case "eth_sign": {
+                if (!this.signer) {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts:102]:
```suggestion
                if (this.signer == null) {
```

File: packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+                const req = ethers.providers.JsonRpcProvider.hexlifyTransaction(params![0]);
+                const result = await this.provider.estimateGas(req);
+                return result.toHexString();
+            }
+
+            // @TOOD: Transform? No uncles?
+            case "eth_getBlockByHash":
+            case "eth_getBlockByNumber": {
+                if (params![1]) {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts:88]:
```suggestion
                if (params![1] != null) {
```

File: packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+            case "eth_sendRawTransaction": {
+                return await this.provider.sendTransaction(params![0]);
+            }
+            case "eth_call": {
+                delete params![0].from;
+                const req = ethers.providers.JsonRpcProvider.hexlifyTransaction(params![0]);
+                return await this.provider.call(req, params![1]);
+            }
+            case "estimateGas": {
+                if (params![1] && params![1] !== "latest") {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts:76]:
```suggestion
                if (params![1] != null && params![1] !== "latest") {
```

File: packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
+import { ZeroDevAccountSigner, ZeroDevEthersProvider } from "@zerodev/sdk";
+import { ethers } from "ethers";
+export class CustomEip1193Bridge {
+    readonly signer: ZeroDevAccountSigner<"ECDSA">;
+    readonly provider: ZeroDevEthersProvider<"ECDSA">;
+
+    constructor(signer:ZeroDevAccountSigner<"ECDSA">, provider: ZeroDevEthersProvider<"ECDSA">) {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts:7]: prettier not running it seems
```suggestion
    constructor(signer: ZeroDevAccountSigner<"ECDSA">, provider: ZeroDevEthersProvider<"ECDSA">) {
```

File: packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+            case "eth_newPendingTransactionFilter":
+            case "eth_uninstallFilter":
+            case "eth_getFilterChanges":
+            case "eth_getFilterLogs":
+            case "eth_getLogs":
+                break;
+        }
+
+        // If our provider supports send, maybe it can do a better job?
+        if ((<any>(this.provider)).send) {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts:178]: what's this? `ZeroDevEthersProvider<"ECDSA">` always has the `send` method defined?

File: packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+            const result = await (<any>(this.provider)).send(method, params);
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts:179]: the `params` error should be resolved with `params ?? []`
```suggestion
            const result = await this.provider.send(method, params ?? []);
```

File: packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+                req.gasLimit = req.gas;
+                delete req.gas;
+                const tx = await this.signer.sendTransaction(req);
+                return tx.hash;
+            }
+
+            case "eth_getUncleCountByBlockHash":
+            case "eth_getUncleCountByBlockNumber":
+            {
+                coerce = ethers.utils.hexValue;
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts:159]: can we comment why this is here and not for other methods?

File: packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts [hunk 10]
```diff
@@ -0,0 +1,186 @@
+import { ZeroDevAccountSigner, ZeroDevEthersProvider } from "@zerodev/sdk";
+import { ethers } from "ethers";
+export class CustomEip1193Bridge {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/CustomBridge.ts:3]: nit: class rename to `ZeroDevEip1193Bridge`, and file name updated to match

File: packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts [hunk 11]
```diff
@@ -35,6 +38,31 @@ export class EVMAAWallet<B extends EVMBlockchainIncludingTestnet = EVMBlockchain
         this.chain = chain;
     }
 
+    async getSigner(type: SignerType) {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts:41]: we need to define the return type of this method such that if you provide `"ethers"`, typescript knows that you're returned an `ethers.Signer`, and if `"viem"`, the return type is `viem.WalletClient`

It'll look something like this:
```
type SignerMap = {
    ethers: ethers.Signer;
    viem: WalletClient;
};

async getSigner<Type extends SignerType = SignerType>(type: Type): Promise<SignerMap[Type]> {
```
but I'm having trouble nailing this down right now, have asked Maxwell for some help

**Review by afeight — COMMENTED** [review]
ok this is mostly good. 
Only blocking comment is that around the retrn type of `getSigner`, which I'm pinging Maxwell on

File: packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts (hunk 11 above)
Comment (afeight) [reply packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts:41]: ah, for now we can do:
```
type SignerMap = {
    ethers: ethers.Signer;
    viem: WalletClient;
};

export class EVMAAWallet<B extends EVMBlockchainIncludingTestnet = EVMBlockchainIncludingTestnet> extends BaseWallet {
    private sessionKeySignerAddress?: string;
    chain: B;

    constructor(provider: ZeroDevEthersProvider<"ECDSA">, crossmintService: CrossmintService, chain: B) {
        super(provider, crossmintService);
        this.chain = chain;
    }

    async getSigner<Type extends SignerType = SignerType>(type: Type): Promise<SignerMap[Type]> {
        switch (type) {
            case "ethers":
                return this.provider.getSigner() as any;
            case "viem":
                const customeip1193Provider = new CustomEip1193Bridge(this.provider.getAccountSigner(), this.provider);
                const customTransport = custom({
                    async request({ method, params }) {
                        return customeip1193Provider.send(method, params);
                    },
                });
                const walletClient: viem.WalletClient = createWalletClient({
                    account: await this.getAddress(),
                    chain: getViemNetwork(this.chain),
                    transport: customTransport,
                }).extend(publicActions);
                return walletClient as any;
            default:
                logError("[GET_SIGNER] - ERROR", {
                    error: errorToJSON("Invalid signer type"),
                });
                throw new Error("Invalid signer type");
        }
    }

**Review by afeight — APPROVED** [review]

**Pull request merged**